logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb

```diff
@@ -6,130 +6,234 @@
+        number_of_times_throttled = -1
+        time_throttled_nanos = -1
 
         lines.each do |line|
           fields = line.split(/\s+/)
           case fields.first
           when "nr_periods" then number_of_elapsed_periods = fields[1].to_i
-          when "nr_throttled" then number_of_times_throttled= fields[1].to_i
+          when "nr_throttled" then number_of_times_throttled = fields[1].to_i
           when "throttled_time" then time_throttled_nanos = fields[1].to_i
           end
         end
 
         CpuStats.new(number_of_elapsed_periods, number_of_times_throttled, time_throttled_nanos)
       end
+    end
 
-      def get_all
-       groups = control_groups
-       return if groups.empty?
+    class UnhandledResource
+      include LogStash::Util::Loggable
+      include ControllerResource
+    end
 
-       cgroups_stats = {
-         :cpuacct => {},
-         :cpu => {}
-       }
+    class CpuStats
+      attr_reader :number_of_elapsed_periods, :number_of_times_throttled, :time_throttled_nanos
 
-       cpuacct_group = groups[GROUP_CPUACCT]
-       cgroups_stats[:cpuacct][:control_group] = cpuacct_group
-       cgroups_stats[:cpuacct][:usage_nanos] = cgroup_cpuacct_usage_nanos(cpuacct_group)
+      def initialize(number_of_elapsed_periods, number_of_times_throttled, time_throttled_nanos)
+        @number_of_elapsed_periods = number_of_elapsed_periods
+        @number_of_times_throttled = number_of_times_throttled
+        @time_throttled_nanos = time_throttled_nanos
+      end
+    end
 
-       cpu_group = groups[GROUP_CPU]
-       cgroups_stats[:cpu][:control_group] = cpu_group
-       cgroups_stats[:cpu][:cfs_period_micros] = cgroup_cpu_fs_period_micros(cpu_group)
-       cgroups_stats[:cpu][:cfs_quota_micros] = cgroup_cpu_fs_quota_micros(cpu_group)
+    class << self
+      def get_all(resources = CGroupResources.new)
+        unless resources.cgroup_available?
+          files = [resources.cgroup_file, resources.cpu_dir, resources.cpuacct_dir]
+          logger.debug("One or more required cgroup files or directories not found: #{files.join(', ')}")
+          return
+        end
 
-       cpu_stats = cgroup_cpuacct_cpu_stat(cpu_group)
+        groups = resources.controller_groups
 
-       cgroups_stats[:cpu][:stat] = {
-         :number_of_elapsed_periods => cpu_stats.number_of_elapsed_periods,
-         :number_of_times_throttled => cpu_stats.number_of_times_throttled,
-         :time_throttled_nanos => cpu_stats.time_throttled_nanos
-       }
+        if groups.empty?
+          logger.debug("The main cgroup file did not have any controllers: #{resources.cgroup_file}")
+          return
+        end
 
-       cgroups_stats
+        cpuacct_controller = groups[resources.controller_cpuacct]
+        cpu_controller = groups[resources.controller_cpu]
+
+        cgroups_stats = {
+         :cpuacct => {},
+         :cpu => {}
+        }
+        cpuacct_controller.fill(cgroups_stats[:cpuacct])
```

Comment:
I would prefer inverting the `#fill` method (which mutates a passed hash) into a `#to_hash` method (that takes no argument and returns a hash), especially since it is only used here, which could easily be refactored to be the following without creating any temporary objects:

~~~ ruby
cgroups_stats = {
  :cpuacct => cpuacct_controller.to_hash,
  :cpu     => cpu_controller.to_hash,
}

~~~

I believe this would also simplify our unit tests, which currently have to juggle a bit to validate side-effects  (e.g. providing empty hash to inject data into).